pkg/scheduler/predicates/ha.go

```diff
@@ -15,117 +15,111 @@ package predicates
 	h := &ha{
 		kubeCli: kubeCli,
+		cli:     cli,
 	}
 	h.podListFn = h.realPodListFn
+	h.pvcGetFn = h.realPVCGetFn
+	h.tcGetFn = h.realTCGetFn
 	return h
 }
 
 func (h *ha) Name() string {
 	return "HighAvailability"
 }
 
-// 1. First, we sort all the nodes we get from kube-scheduler by how many same kind of pod it contains,
-//    find the nodes that have least pods.
-// 2. When scheduling the first replicas pods, we must ensure no previous pods on the nodes.
-// 3. For later pods, we choose the nodes that have least pods.
+// 1. return the node to kube-scheduler if there is only one node and the pod's pvc is bound
+// 2. return these nodes that have least pods and its pods count is less than replicas/2 to kube-scheduler
+// 3. let kube-scheduler to make the final decision
 func (h *ha) Filter(instanceName string, pod *apiv1.Pod, nodes []apiv1.Node) ([]apiv1.Node, error) {
 	ns := pod.GetNamespace()
 	podName := pod.GetName()
+	component := pod.Labels[label.ComponentLabelKey]
+	tcName := getTCNameFromPod(pod, component)
 
-	var component string
-	var exist bool
-	if component, exist = pod.Labels[label.ComponentLabelKey]; !exist {
-		return nodes, fmt.Errorf("can't find component in pod labels: %s/%s", ns, podName)
+	if len(nodes) == 1 {
+		pvcName := fmt.Sprintf("%s-%s", component, podName)
+		pvc, err := h.pvcGetFn(ns, pvcName)
+		if err != nil {
+			return nil, err
+		}
+		if pvc.Status.Phase == apiv1.ClaimBound {
+			return nodes, nil
+		}
 	}
+
 	podList, err := h.podListFn(ns, instanceName, component)
 	if err != nil {
 		return nil, err
 	}
-
-	ordinal, err := util.GetOrdinalFromPodName(podName)
+	tc, err := h.tcGetFn(ns, tcName)
 	if err != nil {
 		return nil, err
 	}
-
-	// when a deleted pod is recreated again, it should be rescheduled to the original node
-	if len(nodes) == 1 {
-		nextPodName := util.GetNextOrdinalPodName(podName, ordinal)
-		for _, pod := range podList.Items {
-			if pod.GetName() == nextPodName && pod.Spec.NodeName != "" {
-				return nodes, nil
-			}
-		}
-	}
+	replicas := getReplicasFrom(tc, component)
 
 	nodeMap := make(map[string][]string)
 	for _, node := range nodes {
 		nodeMap[node.GetName()] = make([]string, 0)
 	}
 	for _, pod := range podList.Items {
-		podName1 := pod.GetName()
+		pName := pod.GetName()
 		nodeName := pod.Spec.NodeName
-		ordinal1, err := util.GetOrdinalFromPodName(podName1)
-		if err != nil {
-			return nil, err
-		}
-
-		if ordinal1 < ordinal && nodeName == "" {
-			return nil, fmt.Errorf("waiting for pod: %s/%s to be scheduled", ns, podName1)
-		}
 		if nodeName == "" || nodeMap[nodeName] == nil {
 			continue
 		}
 
-		nodeMap[nodeName] = append(nodeMap[nodeName], podName1)
+		nodeMap[nodeName] = append(nodeMap[nodeName], pName)
 	}
 	glog.V(4).Infof("nodeMap: %+v", nodeMap)
 
-	var min int
-	var minInitialized bool
-	for _, podNameArr := range nodeMap {
-		count := len(podNameArr)
-		if !minInitialized {
-			minInitialized = true
-			min = count
+	min := -1
```

Comment:
Is it better to initialize the min variable to a large enough value?

Reply:
Large enough, how big is it?

Reply:
The default number of pods that can be run on a kubelet is 110, so I think it's okay to be larger than this number, what do you think?

Reply:
This number may be modified by the user. -1 is good enough i think.